From 23df766413de659d8383ac264bcee80e0570afa7 Mon Sep 17 00:00:00 2001
From: Ulf Hermann <ulf.hermann@qt.io>
Date: Tue, 23 Sep 2025 13:59:52 +0200
Subject: [PATCH 311/553] Core: Decouple QIterable from QMetaContainer

Technically, the thing we're iterating over doesn't have to be a
QMetaContainer. It's templated all the way. Only in one place we
accidentally referenced QMetaContainer and that should be just "auto".

Task-number: QTBUG-140181
Pick-to: 6.8
Change-Id: Ic6b129822345f9ef4c9a117eb4a3e84cbcc148c3
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit b096619acc65b3adb0e6ed1f21296cc430113439)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/kernel/qiterable.h | 9 +++++++--
 src/corelib/kernel/qmetatype.h | 1 +
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/corelib/kernel/qiterable.h b/src/corelib/kernel/qiterable.h
index 4adcdfd76fb..5e25dd1c0a5 100644
--- a/src/corelib/kernel/qiterable.h
+++ b/src/corelib/kernel/qiterable.h
@@ -6,9 +6,12 @@
 
 #include <QtCore/qglobal.h>
 #include <QtCore/qtypeinfo.h>
-#include <QtCore/qmetacontainer.h>
 #include <QtCore/qtaggedpointer.h>
 
+#if !defined(QT_LEAN_HEADERS) || QT_LEAN_HEADERS < 1
+#   include <QtCore/qmetacontainer.h>
+#endif
+
 QT_BEGIN_NAMESPACE
 
 namespace QtPrivate {
@@ -70,7 +73,7 @@ public:
     using iterator_category = IteratorCategory;
     QTaggedIterator(Iterator &&it) : Iterator(std::move(it))
     {
-        const QMetaContainer metaContainer = this->metaContainer();
+        const auto metaContainer = this->metaContainer();
         if constexpr (std::is_base_of_v<std::random_access_iterator_tag, IteratorCategory>) {
             if (!metaContainer.hasRandomAccessIterator()) {
                 qFatal("You cannot use this iterator as a random access iterator");
@@ -98,6 +101,8 @@ public:
                 this->clearIterator();
             }
         }
+
+        Q_UNUSED(metaContainer); // in case none of the above apply
     }
 
     bool operator==(const QTaggedIterator &o) const { return Iterator::operator==(o); }
diff --git a/src/corelib/kernel/qmetatype.h b/src/corelib/kernel/qmetatype.h
index 9bd08ed2e44..921243d580b 100644
--- a/src/corelib/kernel/qmetatype.h
+++ b/src/corelib/kernel/qmetatype.h
@@ -15,6 +15,7 @@
 #include <QtCore/qfloat16.h>
 #include <QtCore/qhashfunctions.h>
 #include <QtCore/qiterable.h>
+#include <QtCore/qmetacontainer.h>
 #ifndef QT_NO_QOBJECT
 #include <QtCore/qobjectdefs.h>
 #endif
-- 
2.51.2

